src/openforms/submissions/form_logic.py

```diff
@@ -214,15 +208,19 @@ def check_submission_logic(
         submission_variables_state.set_values(unsaved_data)
     data_container = DataContainer(state=submission_variables_state)
 
-    mutation_operations = []
+    mutation_operations: list[ActionOperation] = []
     for operation in iter_evaluate_rules(rules, data_container, submission):
```

Comment:
Probably this can be built using a list comprehension now

Reply:
I think a list comprehension makes the code look pure, but it isn't, data mutations are applied as the logic rules are evaluated by looping through them.